### ISSUE TITLE
Implement Machine Learning Model Deployment for Signal Prediction

### ISSUE THREAD

**Issue by Mathews-25** [AgesEmpire/StellarSwipe-Contract#110]
Context:
Deploy trained ML models on-chain for real-time signal prediction and trade automation. Models predict price movements or signal success probability.
Objective:
Build infrastructure to deploy and execute ML models within smart contracts for automated trading decisions.
Scope:

Support lightweight ML models (decision trees, linear models)
Deploy model weights and parameters on-chain
Execute inference in smart contract
Update models periodically
Track model performance and drift

Implementation Guidelines:

Key Files: /contracts/auto_trade/src/strategies/ml_prediction.rs
ML infrastructure:
struct MLTradingStrategy {
      user: Address,
      asset_pair: AssetPair,
      model: MLModel,
      feature_config: FeatureConfig,
      prediction_threshold: u32, // Min prediction confidence to trade
      position_size_pct: u32,
      active_position: Option<MLPosition>,
      model_version: u32,
      last_model_update: u64,
  }
  
  enum MLModel {
      LogisticRegression {
          weights: Vec<i128>,
          intercept: i128,
      },
      DecisionTree {
          nodes: Vec<TreeNode>,
      },
      RandomForest {
          trees: Vec<DecisionTree>,
          tree_weights: Vec<u32>,
      },
      NeuralNetwork {
          layers: Vec<Layer>,
      },
  }
  
  struct TreeNode {
      feature_index: u32,
      threshold: i128,
      left_child: Option<u32>,
      right_child: Option<u32>,
      leaf_value: Option<i128>, // Prediction if leaf node
  }
  
  struct Layer {
      weights: Vec<Vec<i128>>,
      biases: Vec<i128>,
      activation: ActivationFunction,
  }
  
  enum ActivationFunction {
      ReLU,
      Sigmoid,
      Tanh,
      Linear,
  }
  
  struct FeatureConfig {
      features: Vec<FeatureType>,
      normalization: Map<String, NormalizationParams>,
  }
  
  enum FeatureType {
      Price,
      Volume,
      RSI,
      MACD,
      BollingerBands,
      PriceChange { periods: u32 },
      VolumeChange { periods: u32 },
      Custom { name: String, calculation: String },
  }
  
  struct MLPosition {
      position_id: u64,
      predicted_direction: TradeDirection,
      prediction_confidence: u32,
      entry_price: i128,
      amount: i128,
  }
Feature Extraction:
rustfn extract_features(
    asset_pair: AssetPair,
    feature_config: &FeatureConfig
) -> Result<Vec<i128>, Error> {
    let mut features = Vec::new();
    
    for feature_type in &feature_config.features {
        let value = match feature_type {
            FeatureType::Price => {
                get_current_price(asset_pair)?
            },
            FeatureType::Volume => {
                get_24h_volume(asset_pair)?
            },
            FeatureType::RSI => {
                calculate_rsi(asset_pair, 14)? as i128
            },
            FeatureType::MACD => {
                let (macd, signal) = calculate_macd(asset_pair)?;
                macd - signal
            },
            FeatureType::BollingerBands => {
                let (upper, lower, middle) = calculate_bollinger_bands(asset_pair, 20, 2)?;
                let current_price = get_current_price(asset_pair)?;
                // Position within bands
                ((current_price - lower) * 10000) / (upper - lower)
            },
            FeatureType::PriceChange { periods } => {
                let current = get_current_price(asset_pair)?;
                let past = get_price_n_periods_ago(asset_pair, *periods)?;
                ((current - past) * 10000) / past
            },
            FeatureType::VolumeChange { periods } => {
                let current = get_current_volume(asset_pair)?;
                let past = get_volume_n_periods_ago(asset_pair, *periods)?;
                ((current - past) * 10000) / past
            },
            FeatureType::Custom { name, calculation } => {
                calculate_custom_feature(asset_pair, name, calculation)?
            },
        };
        
        // Normalize feature
        let normalized = normalize_feature(value, feature_type, feature_config)?;
        features.push(normalized);
    }
    
    Ok(features)
}

fn normalize_feature(
    value: i128,
    feature_type: &FeatureType,
    config: &FeatureConfig
) -> Result<i128, Error> {
    let feature_name = format!("{:?}", feature_type);
    
    if let Some(params) = config.normalization.get(&feature_name) {
        // Min-max normalization: (x - min) / (max - min)
        let normalized = ((value - params.min) * PRECISION) / (params.max - params.min);
        Ok(clamp(normalized, 0, PRECISION))
    } else {
        Ok(value)
    }
}
Model Inference:
rustfn predict_with_model(
    model: &MLModel,
    features: &[i128]
) -> Result<MLPrediction, Error> {
    match model {
        MLModel::LogisticRegression { weights, intercept } => {
            predict_logistic_regression(features, weights, *intercept)
        },
        MLModel::DecisionTree { nodes } => {
            predict_decision_tree(features, nodes)
        },
        MLModel::RandomForest { trees, tree_weights } => {
            predict_random_forest(features, trees, tree_weights)
        },
        MLModel::NeuralNetwork { layers } => {
            predict_neural_network(features, layers)
        },
    }
}

fn predict_logistic_regression(
    features: &[i128],
    weights: &[i128],
    intercept: i128
) -> Result<MLPrediction, Error> {
    require!(features.len() == weights.len(), "Feature dimension mismatch");
    
    // Calculate dot product
    let mut logit = intercept;
    for i in 0..features.len() {
        logit += (features[i] * weights[i]) / PRECISION;
    }
    
    // Apply sigmoid function
    let probability = sigmoid(logit);
    
    Ok(MLPrediction {
        direction: if probability > 5000 {
            TradeDirection::Buy
        } else {
            TradeDirection::Sell
        },
        confidence: if probability > 5000 {
            probability
        } else {
            10000 - probability
        },
        raw_score: logit,
    })
}

fn sigmoid(x: i128) -> u32 {
    // Sigmoid: 1 / (1 + e^-x)
    // Approximation for on-chain efficiency
    
    if x >= 6 * PRECISION {
        return 10000; // ~1.0
    }
    if x <= -6 * PRECISION {
        return 0; // ~0.0
    }
    
    // Piecewise linear approximation
    let scaled_x = x / 1000; // Scale down for calculation
    let approximation = 5000 + (scaled_x * 833 / PRECISION);
    
    clamp(approximation, 0, 10000) as u32
}

fn predict_decision_tree(
    features: &[i128],
    nodes: &[TreeNode]
) -> Result<MLPrediction, Error> {
    let mut current_node_idx = 0u32;
    
    loop {
        let node = &nodes[current_node_idx as usize];
        
        // Check if leaf node
        if let Some(leaf_value) = node.leaf_value {
            return Ok(MLPrediction {
                direction: if leaf_value > 0 {
                    TradeDirection::Buy
                } else {
                    TradeDirection::Sell
                },
                confidence: min(10000, leaf_value.abs() as u32),
                raw_score: leaf_value,
            });
        }
        
        // Internal node - traverse based on feature value
        let feature_value = features[node.feature_index as usize];
        
        current_node_idx = if feature_value <= node.threshold {
            node.left_child.ok_or(Error::InvalidTreeStructure)?
        } else {
            node.right_child.ok_or(Error::InvalidTreeStructure)?
        };
    }
}

fn predict_random_forest(
    features: &[i128],
    trees: &[DecisionTree],
    tree_weights: &[u32]
) -> Result<MLPrediction, Error> {
    let mut weighted_sum = 0i128;
    let mut total_weight = 0u32;
    
    for (i, tree) in trees.iter().enumerate() {
        let prediction = predict_decision_tree(features, &tree.nodes)?;
        weighted_sum += prediction.raw_score * tree_weights[i] as i128;
        total_weight += tree_weights[i];
    }
    
    let avg_score = weighted_sum / total_weight as i128;
    
    Ok(MLPrediction {
        direction: if avg_score > 0 {
            TradeDirection::Buy
        } else {
            TradeDirection::Sell
        },
        confidence: min(10000, avg_score.abs() as u32),
        raw_score: avg_score,
    })
}

fn predict_neural_network(
    features: &[i128],
    layers: &[Layer]
) -> Result<MLPrediction, Error> {
    let mut activations = features.to_vec();
    
    // Forward pass through each layer
    for layer in layers {
        activations = forward_layer(&activations, layer)?;
    }
    
    // Final layer output
    let output = activations[0];
    
    Ok(MLPrediction {
        direction: if output > 0 {
            TradeDirection::Buy
        } else {
            TradeDirection::Sell
        },
        confidence: min(10000, output.abs() as u32),
        raw_score: output,
    })
}

fn forward_layer(
    inputs: &[i128],
    layer: &Layer
) -> Result<Vec<i128>, Error> {
    let mut outputs = Vec::new();
    
    for (neuron_weights, bias) in layer.weights.iter().zip(&layer.biases) {
        // Weighted sum
        let mut sum = *bias;
        for (input, weight) in inputs.iter().zip(neuron_weights) {
            sum += (input * weight) / PRECISION;
        }
        
        // Apply activation function
        let activated = match layer.activation {
            ActivationFunction::ReLU => max(0, sum),
            ActivationFunction::Sigmoid => sigmoid(sum) as i128,
            ActivationFunction::Tanh => tanh_approx(sum),
            ActivationFunction::Linear => sum,
        };
        
        outputs.push(activated);
    }
    
    Ok(outputs)
}

fn tanh_approx(x: i128) -> i128 {
    // tanh approximation: x / (1 + |x|)
    let abs_x = x.abs();
    let denominator = PRECISION + abs_x;
    
    if denominator == 0 {
        return 0;
    }
    
    (x * PRECISION) / denominator
}
Generate Trading Signal:
rustfn check_ml_signal(strategy_id: u64) -> Result<Option<MLSignal>, Error> {
    let strategy = get_ml_trading_strategy(strategy_id)?;
    
    if strategy.active_position.is_some() {
        return Ok(None);
    }
    
    // Extract features
    let features = extract_features(strategy.asset_pair, &strategy.feature_config)?;
    
    // Make prediction
    let prediction = predict_with_model(&strategy.model, &features)?;
    
    // Check if confidence exceeds threshold
    if prediction.confidence < strategy.prediction_threshold {
        return Ok(None);
    }
    
    Ok(Some(MLSignal {
        direction: prediction.direction,
        confidence: prediction.confidence,
        features: features,
        model_version: strategy.model_version,
    }))
}

fn execute_ml_trade(
    strategy_id: u64,
    signal: MLSignal
) -> Result<u64, Error> {
    let strategy = get_ml_trading_strategy_mut(strategy_id)?;
    
    let portfolio_value = get_portfolio_value(strategy.user)?;
    
    // Scale position size by confidence
    let base_size = (portfolio_value * strategy.position_size_pct as i128) / 10000;
    let confidence_scaled = (base_size * signal.confidence as i128) / 10000;
    
    let trade_id = execute_trade(
        strategy.user,
        strategy.asset_pair,
        confidence_scaled,
        OrderType::Market
    )?;
    
    let position_id = get_next_position_id();
    let current_price = get_current_price(strategy.asset_pair)?;
    
    let position = MLPosition {
        position_id,
        predicted_direction: signal.direction,
        prediction_confidence: signal.confidence,
        entry_price: current_price,
        amount: confidence_scaled,
    };
    
    strategy.active_position = Some(position);
    
    emit_event(MLTradeExecuted {
        strategy_id,
        position_id,
        model_version: signal.model_version,
        confidence: signal.confidence,
    });
    
    Ok(position_id)
}
Model Updates:
rustfn update_ml_model(
    strategy_id: u64,
    new_model: MLModel,
    new_version: u32
) -> Result<(), Error> {
    let strategy = get_ml_trading_strategy_mut(strategy_id)?;
    
    require!(
        new_version > strategy.model_version,
        "Model version must increase"
    );
    
    // Validate model structure
    validate_model_structure(&new_model)?;
    
    strategy.model = new_model;
    strategy.model_version = new_version;
    strategy.last_model_update = current_time();
    
    emit_event(MLModelUpdated {
        strategy_id,
        new_version,
    });
    
    Ok(())
}

fn validate_model_structure(model: &MLModel) -> Result<(), Error> {
    match model {
        MLModel::LogisticRegression { weights, intercept } => {
            require!(weights.len() > 0 && weights.len() <= 50, "Invalid weight count");
            require!(intercept.abs() < i128::MAX / 2, "Invalid intercept");
        },
        MLModel::DecisionTree { nodes } => {
            require!(nodes.len() > 0 && nodes.len() <= 1000, "Invalid tree size");
            // Validate tree structure
            validate_tree_structure(nodes)?;
        },
        MLModel::RandomForest { trees, tree_weights } => {
            require!(trees.len() == tree_weights.len(), "Mismatched tree weights");
            require!(trees.len() <= 100, "Too many trees");
        },
        MLModel::NeuralNetwork { layers } => {
            require!(layers.len() > 0 && layers.len() <= 10, "Invalid layer count");
            // Validate each layer
            for layer in layers {
                require!(layer.weights.len() <= 100, "Layer too large");
            }
        },
    }
    
    Ok(())
}
Performance Tracking:
ruststruct MLModelPerformance {
    model_version: u32,
    total_predictions: u32,
    accurate_predictions: u32,
    false_positives: u32,
    false_negatives: u32,
    avg_confidence_on_correct: u32,
    avg_confidence_on_incorrect: u32,
    sharpe_ratio: i32,
}

fn track_ml_performance(
    strategy_id: u64,
    position: &MLPosition,
    actual_outcome: bool
) -> Result<(), Error> {
    let performance = get_ml_performance_mut(strategy_id)?;
    
    performance.total_predictions += 1;
    
    let predicted_bullish = matches!(position.predicted_direction, TradeDirection::Buy);
    
    if predicted_bullish == actual_outcome {
        performance.accurate_predictions += 1;
        performance.avg_confidence_on_correct = 
            ((performance.avg_confidence_on_correct * (performance.accurate_predictions - 1)) + 
             position.prediction_confidence) / performance.accurate_predictions;
    } else {
        if predicted_bullish {
            performance.false_positives += 1;
        } else {
            performance.false_negatives += 1;
        }
        
        let incorrect_count = performance.total_predictions - performance.accurate_predictions;
        performance.avg_confidence_on_incorrect = 
            ((performance.avg_confidence_on_incorrect * (incorrect_count - 1)) + 
             position.prediction_confidence) / incorrect_count;
    }
    
    // Calculate accuracy
    let accuracy = (performance.accurate_predictions * 10000) / performance.total_predictions;
    
    emit_event(MLPerformanceUpdated {
        strategy_id,
        model_version: performance.model_version,
        accuracy,
    });
    
    Ok(())
}
Model Drift Detection:
rustfn detect_model_drift(strategy_id: u64) -> Result<bool, Error> {
    let performance = get_ml_performance(strategy_id)?;
    
    // Check if recent performance has degraded
    let recent_predictions = get_recent_ml_predictions(strategy_id, 50)?;
    
    if recent_predictions.len() < 20 {
        return Ok(false); // Not enough data
    }
    
    let recent_accuracy = recent_predictions.iter()
        .filter(|p| p.was_correct)
        .count() as u32 * 10000 / recent_predictions.len() as u32;
    
    let historical_accuracy = (performance.accurate_predictions * 10000) / 
                             performance.total_predictions;
    
    // Drift detected if recent accuracy drops >20% from historical
    let drift_threshold = historical_accuracy * 80 / 100;
    
    if recent_accuracy < drift_threshold {
        emit_event(ModelDriftDetected {
            strategy_id,
            historical_accuracy,
            recent_accuracy,
        });
        return Ok(true);
    }
    
    Ok(false)
}
Edge Cases:

Model size too large (storage limits) - use model compression
Inference takes too long (gas limit) - simplify model
Feature data unavailable - use default values or skip prediction
Model overfits - detect via drift monitoring

Definition of Done:

 ML model deployment on-chain
 Feature extraction from market data
 Inference for multiple model types
 Confidence-based position sizing
 Model updates and versioning
 Performance tracking
 Drift detection
 Unit tests verify inference accuracy

Validation:

Deploy logistic regression model
Extract features (RSI, MACD, price change)
Run inference
Verify prediction matches offline calculation
Execute trade based on prediction
Track actual vs predicted outcome
Update model to new version
Detect performance drift
Deploy decision tree model
Test random forest ensemble
